maro/rl/policy/policy_manager.py

```diff
@@ -339,3 +339,193 @@ def exit(self):
+            working directory.
+        proxy_kwargs: Keyword parameters for the internal ``Proxy`` instance. See ``Proxy`` class
+            for details. Defaults to the empty dictionary.
+    """
+    def __init__(
+        self,
+        policy_dict: Dict[str, AbsCorePolicy],
+        group: str,
+        num_trainers: int,
+        update_trigger: Dict[str, int] = None,
+        warmup: Dict[str, int] = None,
+        post_update: Callable = None,
+        log_dir: str = getcwd(),
+        proxy_kwargs: dict = {}
+    ):
+        super().__init__(policy_dict, update_trigger=update_trigger, warmup=warmup, post_update=post_update)
+        peers = {"trainer": num_trainers}
+        self._proxy = Proxy(group, "policy_manager", peers, component_name="POLICY_MANAGER", **proxy_kwargs)
+
+        self._policy2trainer = defaultdict(list)
+        self._trainer2policies = defaultdict(list)
+        self._exp_cache = defaultdict(ExperienceSet)
+        self._num_experiences_by_policy = defaultdict(int)
+        self.num_trainers = num_trainers
+
+        self._logger = Logger("MULTINODE_POLICY_MANAGER", dump_folder=log_dir)
+
+        self.allocate_trainers()
+
+    def allocate_trainers(self):
+        self._policy2trainer = defaultdict(list)
+        self._trainer2policies = defaultdict(list)
+
+        # initialize
+        if len(self._num_experiences_by_policy) == 0:
+            if len(self.policy_dict) >= self.num_trainers:
+                for i, name in enumerate(self.policy_dict):
+                    trainer_id = i % self.num_trainers
+                    self._policy2trainer[name].append(f"TRAINER.{trainer_id}")
+                    self._trainer2policies[f"TRAINER.{trainer_id}"].append(name)
+            else:
+                trainer_id_list = list(range(self.num_trainers))
+                for i, name in enumerate(self.policy_dict):
+                    for trainer_id in trainer_id_list[i::len(self.policy_dict)]:
+                        self._policy2trainer[name].append(f"TRAINER.{trainer_id}")
+                        self._trainer2policies[f"TRAINER.{trainer_id}"].append(name)
+
+        # allocate trainers according to historical experience numbers.
+        else:
+            total_num_policy = sum(self._num_experiences_by_policy.values())
+            average_payload = total_num_policy / self.num_trainers
+
+            offset = 0
+            policy_quota = dict()
+            for name, num_exp in self._num_experiences_by_policy.items():
+                quota = num_exp / average_payload
+                quota = max(1, int(round(quota)))
+                policy_quota[name] = quota
+
+            # adjust quota if any redundancy occurs.
+            redundancy = self.num_trainers - sum(policy_quota.values())
+            if redundancy > 0:
+                busiest_policy = max(policy_quota, key=lambda name: policy_quota[name])
+                policy_quota[busiest_policy] += redundancy
```

Comment:
so only assign the redundant trainer for the top 1 policy?

Reply:
In my opinion it will be most balanced on average. Basically the redundant computing power will be divided into `max(policy_quota.values())` parts and it would cause less imbalance to assignment.

Reply:
[`new place in file differences`](https://github.com/microsoft/maro/pull/377/files#diff-99fede4365e091829ad935775a32eebeac22d29953184d9107e73941373428c5R105)

I think it may be an easy way but not the most balanced way?

Actually, I think allocate the "redundant" workers to the heaviest policy is OK. Just want to confirm your design idea.